Source/DafnyCore/AST/Expressions/Expressions.cs

```diff
@@ -1212,6 +1257,7 @@ public Resolver_IdentifierExpr(IToken tok, TopLevelDecl decl, List<Type> typeArg
     Decl = decl;
     TypeArgs = typeArgs;
     Type = decl is ModuleDecl ? (Type)new ResolverType_Module() : new ResolverType_Type();
+    PreType = decl is ModuleDecl ? new PreTypePlaceholderModule() : new PreTypePlaceholderType();
```

Comment:
Why are we using type-casting `(Type)` in the definition of `Type`, but no equivalent in the definition of `PreType`?

Reply:
Astute observation. It is because neither `ResolverType_Module` nor `ResolverType_Type` is a subclass of the other, whereas `PreTypePlaceholderModule` is a subclass of `PreTypePlaceholderModule`. The rules in C# for inferring the type of a `_ ? _ : _` expression is unable to figure out that both alternatives are a subtype of `Type` on L259.

(Dafny's rules for this are different. In a future PR, you will see that Dafny is willing to infer a pre-type that is common super-type of the two `if-then-else` alternatives. Dafny keeps going up the supertype chain until such a choice is unique. We'll discuss this choice later.)